### PR TITLE
[elastic-agent] Re-introduce instrumentation

### PR DESCRIPTION
**Pull request by stuartnelson3**
cf. https://github.com/elastic/beats/pull/29031 for original description.

seems the issue might have been forgetting to add credentials when starting up the gRPC servers: https://github.com/elastic/beats/commit/76383672ae27867da4c6445e69d9eac604cae335

### CODE REVIEW

File: x-pack/elastic-agent/pkg/agent/application/pipeline/emitter/controller.go
Comment (AndersonQ) [inline x-pack/elastic-agent/pkg/agent/application/pipeline/emitter/controller.go:76]: [Suggestion]
It's safe to call `apm.CaptureError` with a nil error and then call `Send`. Therefore it can be simplified.
```suggestion
		apm.CaptureError(ctx, rErr).Send()
```

File: x-pack/elastic-agent/pkg/agent/application/pipeline/dispatcher/dispatcher.go
Comment (AndersonQ) [inline x-pack/elastic-agent/pkg/agent/application/pipeline/dispatcher/dispatcher.go:84]: [Suggestion]
It's safe to call `apm.CaptureError` with a nil error and then call `Send`. Therefore it can be simplified.
```suggestion
		apm.CaptureError(ctx, err).Send()
```

File: x-pack/elastic-agent/pkg/agent/application/pipeline/emitter/controller.go
Comment (AndersonQ) [inline x-pack/elastic-agent/pkg/agent/application/pipeline/emitter/controller.go:135]: [Suggestion]
It's safe to call `apm.CaptureError` with a nil error and then call `Send`. Therefore it can be simplified.
```suggestion
		apm.CaptureError(ctx, err).Send()
```

File: x-pack/elastic-agent/pkg/agent/application/pipeline/emitter/controller.go
Comment (AndersonQ) [inline x-pack/elastic-agent/pkg/agent/application/pipeline/emitter/controller.go:156]: [Suggestion]
It's safe to call `apm.CaptureError` with a nil error and then call `Send`. Therefore it can be simplified.
```suggestion
		apm.CaptureError(ctx, err).Send()
```

File: x-pack/elastic-agent/pkg/agent/application/pipeline/emitter/emitter.go
Comment (AndersonQ) [inline x-pack/elastic-agent/pkg/agent/application/pipeline/emitter/emitter.go:40]: [Suggestion]
It's safe to call `apm.CaptureError` with a nil error and then call `Send`. Therefore it can be simplified.
```suggestion
		apm.CaptureError(ctx, err).Send()
```

File: x-pack/elastic-agent/pkg/agent/application/pipeline/stream/operator_stream.go
Comment (AndersonQ) [inline x-pack/elastic-agent/pkg/agent/application/pipeline/stream/operator_stream.go:56]: [Suggestion]
It's safe to call `apm.CaptureError` with a nil error and then call `Send`. Therefore it can be simplified.
```suggestion
		apm.CaptureError(ctx, err).Send()
```

File: x-pack/elastic-agent/pkg/agent/application/upgrade/step_download.go
Comment (AndersonQ) [inline x-pack/elastic-agent/pkg/agent/application/upgrade/step_download.go:30]: [Suggestion]
It's safe to call `apm.CaptureError` with a nil error and then call `Send`. Therefore it can be simplified.
```suggestion
		apm.CaptureError(ctx, err).Send()
```

File: x-pack/elastic-agent/pkg/agent/cmd/enroll_cmd.go
Comment (AndersonQ) [inline x-pack/elastic-agent/pkg/agent/cmd/enroll_cmd.go:192]: [Suggestion]
It's safe to call `apm.CaptureError` with a nil error and then call `Send`. Therefore it can be simplified.
```suggestion
		apm.CaptureError(ctx, err).Send()
```

File: x-pack/elastic-agent/pkg/fleetapi/ack_cmd.go
Comment (AndersonQ) [inline x-pack/elastic-agent/pkg/fleetapi/ack_cmd.go:89]: [Suggestion]
It's safe to call `apm.CaptureError` with a nil error and then call `Send`. Therefore it can be simplified.
```suggestion
		apm.CaptureError(ctx, err).Send()
```

File: x-pack/elastic-agent/pkg/fleetapi/acker/fleet/fleet_acker.go
Comment (AndersonQ) [inline x-pack/elastic-agent/pkg/fleetapi/acker/fleet/fleet_acker.go:58]: [Suggestion]
It's safe to call `apm.CaptureError` with a nil error and then call `Send`. Therefore it can be simplified.
```suggestion
		apm.CaptureError(ctx, err).Send()
```

File: x-pack/elastic-agent/pkg/fleetapi/acker/fleet/fleet_acker.go
Comment (AndersonQ) [inline x-pack/elastic-agent/pkg/fleetapi/acker/fleet/fleet_acker.go:86]: [Suggestion]
It's safe to call `apm.CaptureError` with a nil error and then call `Send`. Therefore it can be simplified.
```suggestion
		apm.CaptureError(ctx, err).Send()
```

File: x-pack/elastic-agent/pkg/fleetapi/acker/lazy/lazy_acker.go
Comment (AndersonQ) [inline x-pack/elastic-agent/pkg/fleetapi/acker/lazy/lazy_acker.go:46]: [Suggestion]
It's safe to call `apm.CaptureError` with a nil error and then call `Send`. Therefore it can be simplified.
```suggestion
		apm.CaptureError(ctx, err).Send()
```

File: x-pack/elastic-agent/pkg/fleetapi/acker/lazy/lazy_acker.go
Comment (AndersonQ) [inline x-pack/elastic-agent/pkg/fleetapi/acker/lazy/lazy_acker.go:65]: [Suggestion]
It's safe to call `apm.CaptureError` with a nil error and then call `Send`. Therefore it can be simplified.
```suggestion
		apm.CaptureError(ctx, err).Send()
```

File: x-pack/elastic-agent/pkg/agent/application/pipeline/emitter/controller.go
Comment (AndersonQ) [inline x-pack/elastic-agent/pkg/agent/application/pipeline/emitter/controller.go:146]: [Question]
Any reason to remove information from the error?

File: x-pack/elastic-agent/pkg/agent/application/pipeline/emitter/controller.go
Comment (AndersonQ) [inline x-pack/elastic-agent/pkg/agent/application/pipeline/emitter/controller.go:207]:
```suggestion
	return e.router.Route(ctx, ast.HashStr(), programsToRun)
```

File: x-pack/elastic-agent/pkg/agent/application/pipeline/emitter/emitter.go
Comment (AndersonQ) [inline x-pack/elastic-agent/pkg/agent/application/pipeline/emitter/emitter.go:44]:
```suggestion
		return ctrl.Update(ctx, c)
```

**Review by AndersonQ — CHANGES_REQUESTED** [review]
It isn't critical, but I believe the code would profit from the simplifications I suggested.

File: x-pack/elastic-agent/pkg/agent/application/pipeline/dispatcher/dispatcher.go
Comment (stuartnelson3) [reply x-pack/elastic-agent/pkg/agent/application/pipeline/dispatcher/dispatcher.go:112]: cool, didn't know that

**Comment by stuartnelson3** [comment]
@AndersonQ thanks for the suggestions!

**Comment by axw** [comment]
@AndersonQ is there anything still needed before we can merge this? Stuart is out this week, but I can try and take a look if there's anything remaining.

**Review by AndersonQ — CHANGES_REQUESTED** [review]
Hi @axw 

There are a few things I'd like to ask the reasoning behind or to change:
 - closing the default tracer on the init function
 - wrapping the transport where the transport is declared
 - replacing the use of the `:=` notation by expliciting declaring the variable
 - assigning and error to variable just to return the variable.

I've added the comments on the files.

File: x-pack/elastic-agent/pkg/agent/cmd/enroll_cmd.go
Comment (axw) [inline x-pack/elastic-agent/pkg/agent/cmd/enroll_cmd.go:241]:
```suggestion
			return errors.New("--delay-enroll cannot be used with --fleet-server-es", errors.TypeConfig)
```

File: x-pack/elastic-agent/pkg/agent/cmd/enroll_cmd.go
Comment (axw) [inline x-pack/elastic-agent/pkg/agent/cmd/enroll_cmd.go:249]:
```suggestion
		return errors.New(err, "fail to enroll")
```

File: x-pack/elastic-agent/pkg/agent/cmd/enroll_cmd.go
Comment (axw) [inline x-pack/elastic-agent/pkg/agent/cmd/enroll_cmd.go:256]:
```suggestion
			return errors.New(err, "failed to fix permissions")
```

File: x-pack/elastic-agent/pkg/remote/client.go
Comment (axw) [reply x-pack/elastic-agent/pkg/remote/client.go:129]: This line can just be removed, the transport is already instrumented by `httpcommon.WithAPMHTTPInstrumentation` above. It just never would have done anything before, as trace context wasn't being propagated.

File: x-pack/elastic-agent/pkg/agent/cmd/run.go
Comment (axw) [reply x-pack/elastic-agent/pkg/agent/cmd/run.go:51]: > as apm.DefaultTracer should be closed due to the fact a new tracer is create by NewTracer or NewTracerOptions, better to have them both together. It'll make this relation more explicitly .

The default tracer is not closed because a new one is being created, it's being closed because the default one is not ever expected to be used by Elastic Agent. If it is left open, it will periodically attempt to connect to APM Server. Closing it in package init prevents that as early as possible.

> If in the far future someone else changes the code and decides to use the default tracer, there is a good change they will miss it's being closed here.

We will be releasing a new major version of the agent soon (weeks), and this new version has a more sensible API which does not have side effects due to importing the agent's main package:

```go
package apm

// DefaultTracer returns the default global Tracer, set the first time the
// function is called. It is configured via environment variables.
//
// This will always be initialized to a non-nil value. If any of the
// environment variables are invalid, the corresponding errors will be logged
// to stderr and the default values will be used instead.
func DefaultTracer() *Tracer

// SetDefaultTracer sets the tracer returned by DefaultTracer(). If another
// tracer has already been initialized, it is closed. Any queued events are not
// flushed; it is the responsibility of the caller to call
// DefaultTracer().Flush()
```

I suggest we stick with closing the tracer in package-init for now, and migrate to the new version of the agent ASAP. Then the close call can simply be removed.

File: x-pack/elastic-agent/pkg/agent/application/pipeline/dispatcher/dispatcher.go
Comment (axw) [inline x-pack/elastic-agent/pkg/agent/application/pipeline/dispatcher/dispatcher.go:109]: @AndersonQ @ph I just noticed this while looking through the diff. Is this change safe?

**Comment by axw** [comment]
Thanks @AndersonQ! Agreed on (almost) all points.

File: x-pack/elastic-agent/pkg/agent/application/pipeline/dispatcher/dispatcher.go
Comment (AndersonQ) [reply x-pack/elastic-agent/pkg/agent/application/pipeline/dispatcher/dispatcher.go:109]: We're on an odd scenario here, there are 2 contexts for this method, this shouldn't happen.
 - `ad.ctx` is for cancelation, it shuts the `ActionDispatcher` down
 - `ctx` is just a carrier for the span.
 
Any proper solution would require a good amount of refactoring. Also passing in just the span isn't ideal.

For simplicity, I suggest to document that, let's comment the method and make clear that for now the only allowed function for the context received as parameter is to carry the span, and that cancelation is done through the `ActionDispatcher` context.

I'm adding my suggestion to fix that

@ph, what do you think?

File: x-pack/elastic-agent/pkg/agent/application/pipeline/dispatcher/dispatcher.go
Comment (AndersonQ) [inline x-pack/elastic-agent/pkg/agent/application/pipeline/dispatcher/dispatcher.go:85]: My suggestion to have a context with the `ad.ctx` cancelation and the span. See [this](https://github.com/elastic/beats/pull/30385/files#r813778664) comment for the reasoning behind this.


```suggestion
// ctx is used here ONLY to carry the span, for cancellation use the cancel
// function of the ActionDispatcher.ctx.
func (ad *ActionDispatcher) Dispatch(ctx context.Context, acker store.FleetAcker, actions ...fleetapi.Action) (err error) {
	span, ctx := apm.StartSpan(ctx, "dispatch", "app.internal")
	// Creating a child context that carries both the ad.ctx cancellation and
	// the span from ctx.
	ctx, _ = context.WithCancel(ad.ctx) // ignoring the cancel func as ctx will be cancelled when ever ad.ctx is cancelled.
	ctx = apm.ContextWithSpan(ctx, span)

```

**Review by AndersonQ — CHANGES_REQUESTED** [review]
@axw [this](https://github.com/elastic/beats/pull/30385/files#r813506319) was a good catch. We need to fix that.

I proposed a solution, let me know what you think.

File: x-pack/elastic-agent/pkg/agent/application/pipeline/dispatcher/dispatcher_test.go
Comment (AndersonQ) [inline x-pack/elastic-agent/pkg/agent/application/pipeline/dispatcher/dispatcher_test.go:52]: Given [my change proposal](https://github.com/elastic/beats/pull/30385/files#r813779788), better to add a test.


This is the type definition, I couldn't add as a change suggestion
```
type mockAcker struct {
	CommitFn func(ctx context.Context) error
}

func (m mockAcker) Ack(ctx context.Context, action fleetapi.Action) error {
	panic("implement me")
}

func (m mockAcker) Commit(ctx context.Context) error {
	return m.CommitFn(ctx)
}
```

```suggestion
	ack := noopacker.NewAcker()

	t.Run("Merges ActionDispatcher ctx cancel and Dispatch cxt value", func(t *testing.T) {
		action1 := &mockAction{}
		def := &mockHandler{}

		span := apmtest.NewRecordingTracer().
			StartTransaction("ignore", "ignore").
			StartSpan("ignore", "ignore", nil)

		ctx1, cancel := context.WithCancel(context.Background())
		ack := mockAcker{CommitFn: func(ctx context.Context) error {
			// ctx1 not cancelled yet
			require.NoError(t, ctx.Err())

			got := apm.SpanFromContext(ctx)
			require.Equal(t, span.TraceContext().Span, got.ParentID())

			cancel() // cancel function from ctx1
			require.Error(t, ctx.Err(), context.Canceled)

			return nil
		}}

		d, err := New(ctx1, nil, def)
		require.NoError(t, err)

		ctx2 := apm.ContextWithSpan(context.Background(), span)
		err = d.Dispatch(ctx2, ack, action1)
		require.NoError(t, err)
	})

	t.Run("Success to dispatch multiples events", func(t *testing.T) {
		ctx := context.Background()
```

**Comment by stuartnelson3** [comment]
@AndersonQ i updated the code as per your suggestions, thank you! the one change I made was wrt discarding the `cancel()` func. golint won't let this happen, so I added a `defer cancel()`. I looked through the code and it looked like this should be fine (it seems each ctx except for `ad.ctx` is scoped to the function call?), but let me know if there's a reason we shouldn't be canceling this sub context!

**Review by ph — APPROVED** [review]
This looks ok on my side, I haven't tested it yet.
Just want to make sure this goes in the next minor correct?

**Comment by stuartnelson3** [comment]
@ph Yes, definitely. And if at all possible, getting it merged before elastic-agent is migrated to its own repo and this PR needs to be re-opened there :)

**Comment by ph** [comment]
@AndersonQ Can you do a manual test of this PR so we can unblock @stuartnelson3 ?

File: x-pack/elastic-agent/pkg/agent/application/pipeline/dispatcher/dispatcher.go
Comment (AndersonQ) [inline x-pack/elastic-agent/pkg/agent/application/pipeline/dispatcher/dispatcher.go:82]: [Blocker]
@stuartnelson3 the span isn't being ended anymore or capturing the error. Was it intentional?

File: x-pack/elastic-agent/pkg/agent/application/pipeline/dispatcher/dispatcher.go
Comment (AndersonQ) [inline x-pack/elastic-agent/pkg/agent/application/pipeline/dispatcher/dispatcher.go:86]: If you make changes for the comment above, could you also change here to use `:=`?
Is there a special reason you avoid using `:=`?
```suggestion
	ctx, cancel := context.WithCancel(ad.ctx)
```

**Review by AndersonQ — CHANGES_REQUESTED** [review]

File: x-pack/elastic-agent/pkg/agent/application/pipeline/dispatcher/dispatcher.go
Comment (stuartnelson3) [reply x-pack/elastic-agent/pkg/agent/application/pipeline/dispatcher/dispatcher.go:82]: That was my mistake from cargo-culting your code suggestion above, i'll add it back

File: x-pack/elastic-agent/pkg/agent/application/pipeline/dispatcher/dispatcher.go
Comment (stuartnelson3) [reply x-pack/elastic-agent/pkg/agent/application/pipeline/dispatcher/dispatcher.go:86]: no reason, adding it back

**Comment by stuartnelson3** [comment]
@ph @AndersonQ any idea what the failures are? they only started with https://github.com/elastic/beats/pull/30385/commits/61da5fc812f4c144715f82287cd8153b936d66fd but i have no idea why.

from the e2e failure:

```
[2022-03-02T12:04:01.357Z] tar: /tmp/beats-events-20220302.ndjson: Cannot stat: No such file or directory
[2022-03-02T12:04:01.357Z] tar: Exiting with failure status due to previous errors
```

**Comment by ph** [comment]
@stuartnelson3 I've never seen this on, not sure what `beats-events*.json`, I will check when this CI run is complete.

**Comment by mdelapenya** [comment]
I'd say the error could be related to a network glitch while downloading the elastic-agent from Elastic's artifacts API:

>[2022-03-03T10:59:24.323Z] DEBU[2022-03-03T10:59:21Z] The Elastic artifacts API is available        elapsedTime=141.051415ms retries=1 statusEndpoint="https://artifacts-api.elastic.co/v1/search/8.2.0-fdde08ec/elastic-agent?x-elastic-no-kpi=true"
[2022-03-03T10:59:24.323Z] TRAC[2022-03-03T10:59:21Z] Artifact found                                artifact=elastic-agent artifactName=elastic-agent-8.2.0-SNAPSHOT-linux-arm64.tar.gz elapsedTime=141.343787ms retries=1 version=8.2.0-fdde08ec
[2022-03-03T10:59:24.323Z] TRAC[2022-03-03T10:59:21Z] Downloading file                              path=/tmp/b331ed45-42e7-4c67-97d1-b368749727a7/94807414-5062-4430-9b98-cf6792ba546a url="https://snapshots.elastic.co/8.2.0-fdde08ec/downloads/beats/elastic-agent/elastic-agent-8.2.0-SNAPSHOT-linux-arm64.tar.gz"
[2022-03-03T10:59:24.323Z] TRAC[2022-03-03T10:59:21Z] File downloaded                               elapsedTime=381.276435ms path=/tmp/b331ed45-42e7-4c67-97d1-b368749727a7/94807414-5062-4430-9b98-cf6792ba546a retries=1 url="https://snapshots.elastic.co/8.2.0-fdde08ec/downloads/beats/elastic-agent/elastic-agent-8.2.0-SNAPSHOT-linux-arm64.tar.gz"
[2022-03-03T10:59:28.535Z] ERRO[2022-03-03T10:59:27Z] Could not write file                          error="stream error: stream ID 1; INTERNAL_ERROR" path=/tmp/b331ed45-42e7-4c67-97d1-b368749727a7/94807414-5062-4430-9b98-cf6792ba546a url="https://snapshots.elastic.co/8.2.0-fdde08ec/downloads/beats/elastic-agent/elastic-agent-8.2.0-SNAPSHOT-linux-arm64.tar.gz"
[2022-03-03T10:59:28.535Z] ERRO[2022-03-03T10:59:27Z] Could not download the binary for the Elastic artifact  arch=arm64 artifact=elastic-agent error="stream error: stream ID 1; INTERNAL_ERROR" extension=tar.gz os=linux version=8.2.0-fdde08ec-SNAPSHOT
[2022-03-03T10:59:28.535Z] ERRO[2022-03-03T10:59:27Z] Could not download the binary                 arch=arm64 artifact=elastic-agent error="stream error: stream ID 1; INTERNAL_ERROR" extension=tar.gz os=linux version=8.2.0-fdde08ec-SNAPSHOT

It happens at the very beginning of the execution of that stage, which correlates with the failing scenario (the first one that is executed). After that, further downloads of the binary works.

I'd say that the error is not caused by this PR and could be ignored, as any other execution should work.

**Comment by stuartnelson3** [comment]
@ph @AndersonQ how do you two feel about @mdelapenya 's comment?

this sounds like a transient network error, so I'll re-run to see if it works

**Comment by simitt** [comment]
@ph and @AndersonQ please take some time to support getting this merged in **before** the agent is moved to the new repo. This has been ongoing for quite a while and I think the important part is to get the base merged. Any further details could be ironed out on a follow up; I assume you would like to extend the instrumentation over time anyways.

**Comment by mdelapenya** [comment]
The current e2e test failure is under our radar: https://github.com/elastic/e2e-testing/pull/2203

**Comment by simitt** [comment]
Thanks for the feedback @ph. @stuartnelson3 is on PTO, can you or @AndersonQ please take ownership of merging this in or specifying what exactly is missing?

**Comment by simitt** [comment]
I see that the Elastic Agent has moved to it's own repository. @ph can you please own transfering this PR over and get it over the finish line?

**Comment by stuartnelson3** [comment]
my PTO is being shifted slightly so I'll take a look at transferring this PR over today. I'll catch up with @ph if there is anything I need help with when transferring

**Pull request closed without merging**
